pytest/conftest.py

```diff
@@ -0,0 +1,188 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Default parameters, commandline arguments and common routines for the TensorFlow backend unit tests.
+"""
+# pylint: disable=redefined-outer-name
+import os
+import pytest
+
+from strawberryfields import Engine
+from strawberryfields.backends.base import BaseBackend
+from strawberryfields.backends.fockbackend import FockBackend
+from strawberryfields.backends.gaussianbackend import GaussianBackend
+from strawberryfields.backends.tfbackend import TFBackend
+
+
+# defaults
+TOL = 1e-3
+CUTOFF = 6
+ALPHA = 0.1
+HBAR = 2
+BATCHED = False
+BATCHSIZE = 2
+
+
+@pytest.fixture(scope="session")
+def tol():
+    """Numerical tolerance for equality tests."""
+    return float(os.environ.get("TOL", TOL))
+
+
+@pytest.fixture(scope="session")
+def cutoff():
+    """Fock state cutoff"""
+    return int(os.environ.get("CUTOFF", CUTOFF))
+
+
+@pytest.fixture(scope="session")
+def alpha():
+    """Maximum magnitude of coherent states used in tests"""
+    return float(os.environ.get("ALPHA", ALPHA))
+
+
+@pytest.fixture(scope="session")
+def hbar():
+    """The value of hbar"""
+    return float(os.environ.get("HBAR", HBAR))
+
+
+@pytest.fixture(
+    params=[
+        pytest.param(True, marks=pytest.mark.pure),
+        pytest.param(False, marks=pytest.mark.mixed),
+    ]
+)
+def pure(request):
+    """Whether to run the backend in pure or mixed state mode"""
+    return request.param  # bool(int(os.environ.get("PURE", PURE)))
+
+
+@pytest.fixture(scope="session")
+def batch_size():
+    """Whether to run the backend in batched mode"""
+    if "BATCHSIZE" in os.environ:
+        # if user-specified BATCHSIZE provided, then batching is assumed (even if BATCHED=0)
+        return int(os.environ["BATCHSIZE"])
+
+    # check if batching is turned on
+    batched = bool(int(os.environ.get("BATCHED", BATCHED)))
+
+    if batched:
+        # use the default batch size
+        return BATCHSIZE
+
+    return None  # no batching
+
+
+@pytest.fixture
+def backend(monkeypatch):
+    """Create a mocked out backend fixture for front-end only tests"""
+    dummy_backend = BaseBackend()
+    with monkeypatch.context() as m:
+        # mock out the base backend
+        m.setattr(dummy_backend, "add_mode", lambda n: None)
+        m.setattr(dummy_backend, "del_mode", lambda n: None)
+        m.setattr(dummy_backend, "displacement", lambda alpha, modes: None)
+        m.setattr(dummy_backend, "squeeze", lambda r, modes: None)
+        m.setattr(dummy_backend, "rotation", lambda r, modes: None)
+        m.setattr(dummy_backend, "beamsplitter", lambda t, r, m1, m2: None)
+        m.setattr(dummy_backend, "measure_homodyne", lambda phi, modes, select: 5)
+        m.setattr(dummy_backend, "state", lambda modes: None)
+        m.setattr(dummy_backend, "reset", lambda: None)
+        yield dummy_backend
+
+
+@pytest.fixture(scope="session")
```

Comment:
A fixture that is labelled 'session' is initialized only once, at the beginning of the test suite, rather than intializing before and after every test. This just saves a bit of memory/computational time.